### ISSUE TITLE
Query for attribute :where tuple value always fails

### ISSUE THREAD

**Issue by Quezion** [denistakeda/re-posh#32]
I have a query subscription with a :where clause of `[?e :nsm.entity/form [:todo :nsm.entity/todo-1]]`. It's being transformed into an erroneous query before being passed to DataScript.

Depending on whether I list the value `[:todo :nsm.entity/todo-1]` in the :where clause, or whether I input it as a variable, I get two different errors.

**A) Including the tuple value directly in the :where clause**
```clj
(re-posh/reg-sub
 :todos
 (fn [_ _]
   {:type :query
    :query '[:find ?e
             :in $
             :where
             [?e :nsm.entity/form [:todo :nsm.entity/todo-1]]]}))

Uncaught Error: ?e is not ISeqable
    at Object.cljs$core$seq [as seq] (core.cljs:1226)
    at Object.cljs$core$empty_QMARK_ [as empty_QMARK_] (core.cljs:2115)
    at posh$lib$q_analyze$normalize_all_eavs (q_analyze.cljc:78)
    at core.cljs:4719
    at core.cljs:4719
    at Object.sval (core.cljs:3422)
    at Object.cljs$core$ISeqable$_seq$arity$1 (core.cljs:3479)
    at Object.cljs$core$_seq [as _seq] (core.cljs:716)
    at Object.cljs$core$seq [as seq] (core.cljs:1213)
    at Function.cljs$core$IFn$_invoke$arity$3 (core.cljs:2464)
```

**B) Including tuple value as a :variable**

```clj
(re-posh/reg-sub
 :todos
 (fn [_ _]
   {:type :query
    :query '[:find ?e
             :in $ ?form
             :where
             [?e :nsm.entity/form ?form]]
    :variables [[:todo :nsm.entity/todo-1]]}))

#error {:message 
"Lookup ref attribute … :nsm.entity/todo-1]"
Lookup ref attribute should be marked as :db/unique: [:todo :nsm.entity/todo-1]
, :data {:error :lookup-ref/unique, :entity-id [:todo :nsm.entity/todo-1]}}
datascript$db$entid | @ | db.cljc:833
-- | -- | --
  | (anonymous) | @ | q_analyze.cljc:386
  | (anonymous) | @ | core.cljs:3422
  | (anonymous) | @ | core.cljs:3479
  | cljs$core$_seq | @ | core.cljs:716
  | cljs$core$seq | @ | core.cljs:1213
  | cljs$core$set | @ | core.cljs:9416
  | posh$lib$q_analyze$resolve_any_idents | @ | q_analyze.cljc:385
```
* Same error result if I remove `:variables` from reg-sub & pass it in via `(subscribe [:todos [:todo :nsm.entity/todo-1]])`

**Expected Behavior**

The query should produce the entity-ids of entities in the database that match the tuple in the `:where [?e :nsm.entity/form [...]]`  clause.

**Notes**

* Above query works if I use the raw DataScript API
```clj
(d/q '[:find ?e
           :in $ ?form
           :where
           [?e :nsm.entity/form ?form]]
  (d/db @(get-datascript-db)) [:todo :nsm.entity/todo-1])

#{[3]}
```
* The Schema for this attribute in DataScript & Datomic, respectively:
```clj
;; This is passed to Datascript (as it only supports subset)
:nsm.entity/form {:db/cardinality :db.cardinality/one}

;; datomic (not used, but here for exposition)
{:db/ident :nsm.entity/form,
  :db/valueType :db.type/tuple,
  :db/tupleTypes [:db.type/keyword :db.type/keyword],
  :db/cardinality :db.cardinality/one,
  :db/doc "Tuple [type,spec] of entity"}
```
* In project.clj, I tried `[re-posh "0.3.0" :exclusions [datascript]]` in order to get newer [DataScript 0.18.5](https://github.com/tonsky/datascript#latest-version-) but above errors still occurred

**Comment by Quezion** [comment]
Based off **Scenario B**, it seems that `posh$lib$q_analyze$resolve_any_idents` is the clue. I'll try modifying this -- my hunch is that it's assuming _all_ input vectors are automatically Lookup ref attributes.

As far as historical context, I suspect this error never surfaced because Datomic didn't support tuples until June this year. https://blog.datomic.com/2019/06/tuples-and-database-predicates.html

**Comment by Quezion** [comment]
After investigating the internals, this issue should be instead opened against the Posh repo. I'm going to continue documenting investigation here as I'm not familiar with Posh enough to generate an example against that repo directly. However, I understand if you decide to go ahead & close this @denistakeda 
----------------------

As mentioned before, I modified `posh.lib.q-analyze` from its original definition to ignore Lookup refs entirely.

```clj
;; original
(defn resolve-any-idents [entid-fn db input-set]
    (set (for [x input-set]
           (if (coll? x) (entid-fn db x) x))))

;; modified
(defn resolve-any-idents [entid-fn db input-set]
  (set (for [x input-set]
         (if (coll? x) x x))))
```

After rebuilding posh, re-posh, & my app, I found that the **B** scenario would work (passing the ?form via `:variables`.) The **A** doesn't.

**TODOS:**
- [x] Fix `normalize-all-eavs` to handle tuple values
- [x] Fix `resolve-any-idents` to still work on idents AND tuple values

These fixes may require functions to inspect the transacted schema & notice `:db/valueType tuple`... which isn't currently allowed by DataScript, meaning a possible PR there.)

**Comment by Quezion** [comment]
Per the re-posh PR above:

```
I've published [org.clojars.questyarbrough/posh "0.5.7"] which includes this PR. It's working for my app though it's only exercising small set of DataScript's functionality.
```

Posh hasn't received updates for a few years so I don't know when mpdairy will uptake this PR, but FYI.

**Comment by denistakeda** [comment]
This is fixed in version `0.3.1`
